### PR TITLE
Deprecate `updateUser` GQL mutation in favour of "before update entity" hooks

### PR DESCRIPTION
**Pull request by benwerner01**
## 🌟 What is the purpose of this PR?

This PR deprecates the `updateUser` GQL mutation, by adding the concept of "before update entity" hooks so that the equivalent functionality can be provided by the `updateEntity` mutation.

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1203007126736613/1203251067111511/f) _(internal)_

## 🚫 Blocked by

- https://github.com/hashintel/hash/pull/1280

## 🔍 What does this change?

- removes the `updateUser` GQL query
- adds a "before update entity" hook that performs shortname & preferred name validation on any user entity that is updated with the `updateEntity` GQL query
- adds a `useUpdateAuthenticatedUser` react hook on the frontend to abstract using the `updatePersistedEntity` GQL mutation for updating the logged-in user's preferred name & shortname

## 📜 Does this require a change to the docs?

Not that I'm aware of.

## ⚠️ Known issues

- Signing up currently doesn't automatically redirect you to the "complete signup" page, requiring a refresh

## 🐾 Next steps

- [Deprecate the User GQL type definition](https://app.asana.com/0/1203007126736610/1203244611385365/f)

## 🛡 What tests cover this?

Currently we don't have tests on the GraphQL layer or the frontend.

## ❓ How to test this?

1. Checkout the branch, and start the developer environment
2. Use the existing signup flow to register with an email and a password
3. Refresh the page (currently there's a bug that doesn't correctly redirect to the "complete signup" page
4. Enter a preferred name and shortname, and click submit to test updating a user's preferred name and shortname via the `updateEntity` mutation.

## 📹 Demo

https://user-images.githubusercontent.com/42802102/199074583-a3cbe41b-6901-476d-b873-68253fc5f4f5.mov

### CODE REVIEW

File: packages/hash/frontend/src/components/hooks/useUpdateAuthenticatedUser.ts
Comment (Alfred-Mountfield) [reply packages/hash/frontend/src/components/hooks/useUpdateAuthenticatedUser.ts:48]: As an alternative to passing around the ID and such, you could do:
`getRootsAsEntities(subgraph).pop()`

File: packages/hash/frontend/src/components/hooks/useUpdateAuthenticatedUser.ts
Comment (Alfred-Mountfield) [inline packages/hash/frontend/src/components/hooks/useUpdateAuthenticatedUser.ts:90]: See above comment about

`getRootsAsEntities(subgraph).pop()`

File: packages/hash/frontend/src/components/hooks/useUpdateAuthenticatedUser.ts
Comment (Alfred-Mountfield) [inline packages/hash/frontend/src/components/hooks/useUpdateAuthenticatedUser.ts:98]: Is this safe?

File: packages/hash/frontend/src/pages/signup.page.tsx
Comment (Alfred-Mountfield) [inline packages/hash/frontend/src/pages/signup.page.tsx:222]: This length check is redundant right?

File: packages/hash/frontend/src/components/hooks/useUpdateAuthenticatedUser.ts
Comment (Alfred-Mountfield) [reply packages/hash/frontend/src/components/hooks/useUpdateAuthenticatedUser.ts:97]: This also applies to the code in `useAuthenticatedUser` which I missed in the previous PR review.. sorry

File: packages/hash/frontend/src/components/hooks/useAuthenticatedUser.ts
Comment (Alfred-Mountfield) [inline packages/hash/frontend/src/components/hooks/useAuthenticatedUser.ts:96]: Just commenting here because I can't comment elsewhere on the diff. (And again, apologies I'm only thinking about this now rather than on the previous PR), do we even need the `me` query to do what it does?

Presumably we could use `getEntity` and benefit from any patching or recursive queries, etc. that we've defined for it... I guess we'd still need _a_ `me` query or something to get the appropriate `entityId` of the account entity?

Not suggesting we should actually change this but just a thought

**Review by Alfred-Mountfield — COMMENTED** [review]
Thanks Ben, a few thoughts

File: packages/hash/frontend/src/components/hooks/useUpdateAuthenticatedUser.ts
Comment (luisbettencourt) [inline packages/hash/frontend/src/components/hooks/useUpdateAuthenticatedUser.ts:117]:
```suggestion
  return [
    updateAuthenticatedUser: async (params: UpdateAuthenticatedUserParams) => {
      try {
        setLoading(true);
        return updateAuthenticatedUser(params);
      } finally {
        setLoading(false);
      }
    },
    loading,
  };
};
```

File: packages/hash/frontend/src/components/hooks/useUpdateAuthenticatedUser.ts
Comment (luisbettencourt) [inline packages/hash/frontend/src/components/hooks/useUpdateAuthenticatedUser.ts:116]: The pattern we usually follow for mutation hooks is returning an array with the update function and an object with the loading variable. 

```suggestion
  return [
    async (params: UpdateAuthenticatedUserParams) => {
      try {
        setLoading(true);
        return updateAuthenticatedUser(params);
      } finally {
        setLoading(false);
      }
    },
    { loading },
  ] as const;
```

**Review by luisbettencourt — CHANGES_REQUESTED** [review]

File: packages/hash/api/src/graphql/resolvers/knowledge/entity/before-update-entity-hooks.ts
Comment (benwerner01) [reply packages/hash/api/src/graphql/resolvers/knowledge/entity/before-update-entity-hooks.ts:37]: This would mean the model classes would need to throw GraphQL errors, something we've tried to decouple from the model classes in the past as the model classes aren't always used in environments where GraphQL should be a dependency.

If you feel strongly, we could have the methods throw regular errors, and then catch them in this method to re-throw them as `ApolloError`/`UserInputError` here? But not sure if that's worth the effort.

File: packages/hash/api/src/graphql/resolvers/knowledge/entity/before-update-entity-hooks.ts
Comment (benwerner01) [reply packages/hash/api/src/graphql/resolvers/knowledge/entity/before-update-entity-hooks.ts]: No we can use it here, I just forgot. Added in https://github.com/hashintel/hash/pull/1299/commits/6aa4c09c53a1c91cf1ab419bc201cb9b23c98aea - thanks for the reminder :)

File: packages/hash/frontend/src/components/hooks/useUpdateAuthenticatedUser.ts
Comment (benwerner01) [reply packages/hash/frontend/src/components/hooks/useUpdateAuthenticatedUser.ts:27]: I don't think partial data would be returned here, I chose this error policy so that the consumer of this hook can easily make use of the returned error messages without having to wrap a function call in a `try { ... } catch { ... }`

File: packages/hash/frontend/src/components/hooks/useUpdateAuthenticatedUser.ts
Comment (benwerner01) [reply packages/hash/frontend/src/components/hooks/useUpdateAuthenticatedUser.ts]: Good spot - thanks for letting me know :)

Have addressed in https://github.com/hashintel/hash/pull/1299/commits/df064bd05499fc7b6a909b4ac9fece3857c5f503

File: packages/hash/frontend/src/components/hooks/useUpdateAuthenticatedUser.ts
Comment (benwerner01) [reply packages/hash/frontend/src/components/hooks/useUpdateAuthenticatedUser.ts]: Yes, I think this would be really useful. Have [created the task](https://app.asana.com/0/1202805690238892/1203285029221330/f)  and referenced the todo in https://github.com/hashintel/hash/pull/1299/commits/93decc09336dfed4e55281e66d72dc4e68ef25f0

File: packages/hash/api/src/graphql/resolvers/knowledge/entity/before-update-entity-hooks.ts
Comment (benwerner01) [reply packages/hash/api/src/graphql/resolvers/knowledge/entity/before-update-entity-hooks.ts:49]: This is a good point, I suggest we go with a more robust TS definition of the `properties` object like this:
```TypeScript
type PropertiesObjectValue =
  | string
  | number
  | boolean
  | null
  | object
  | { [propertyTypeBaseUri: BaseUri]: PropertiesObjectValue }
  | PropertiesObjectValue[];

type PropertiesObject = {
  [propertyTypeBaseUri: BaseUri]: PropertiesObjectValue;
};
```

We should probably also use it in the GraphQL type definitions, and replace existing typings of the `properties` with this type. But this should probably be done in a follow up PR. If you agree I'll create a task, and reference a `@todo` here for now?

File: packages/hash/frontend/src/pages/signup.page.tsx
Comment (benwerner01) [reply packages/hash/frontend/src/pages/signup.page.tsx:222]: I'm not sure it is in this case, if there's any circumstance where GQL returns `errors: []` then `parseGraphQLError` will throw an error, so am just trying to account for that. I could move this check into that method though.

File: packages/hash/frontend/src/components/hooks/useUpdateAuthenticatedUser.ts
Comment (benwerner01) [reply packages/hash/frontend/src/components/hooks/useUpdateAuthenticatedUser.ts]: Removed in https://github.com/hashintel/hash/pull/1299/commits/e50e3ae76217ee8a9dfa4a241dc8e0e2260177c7

File: packages/hash/frontend/src/components/hooks/useUpdateAuthenticatedUser.ts
Comment (benwerner01) [reply packages/hash/frontend/src/components/hooks/useUpdateAuthenticatedUser.ts]: I think so, yes - because if `kratosSession` is `undefined` an error would've been thrown in the above queries (as they require the user to be authenticated with a `kratosSession`. Would you like me to add a comment briefly explaining this?

File: packages/hash/api/src/graphql/resolvers/knowledge/entity/before-update-entity-hooks.ts
Comment (Alfred-Mountfield) [reply packages/hash/api/src/graphql/resolvers/knowledge/entity/before-update-entity-hooks.ts:38]: > If you feel strongly, we could have the methods throw regular errors, and then catch them in this method to re-throw them as ApolloError/UserInputError here? But not sure if that's worth the effort.

I think it sounds sensible to do. We need to improve our error handling all over the place it's really not ideal right now. And just saying something like "invalid preferred name" is basically useless without explaining what condition was broken.

File: packages/hash/api/src/graphql/resolvers/knowledge/entity/before-update-entity-hooks.ts
Comment (Alfred-Mountfield) [reply packages/hash/api/src/graphql/resolvers/knowledge/entity/before-update-entity-hooks.ts:49]: Yeah as suggested offline, as a follow-up makes sense

File: packages/hash/frontend/src/components/hooks/useUpdateAuthenticatedUser.ts
Comment (Alfred-Mountfield) [reply packages/hash/frontend/src/components/hooks/useUpdateAuthenticatedUser.ts:27]: This feels like one of those decisions that hasn't been appropriately captured. Do we want an Apollo style guide or something in the repo? Like if we're going to use this here purely to stop a try catch block, surely we want to use it on every mutation? I don't want to be digging around the error policy of the mutation to find out how to catch my errors.

File: packages/hash/frontend/src/components/hooks/useUpdateAuthenticatedUser.ts
Comment (Alfred-Mountfield) [reply packages/hash/frontend/src/components/hooks/useUpdateAuthenticatedUser.ts]: Thanks, just to flag though, I've updated the task to capture what I meant.

File: packages/hash/frontend/src/components/hooks/useUpdateAuthenticatedUser.ts
Comment (Alfred-Mountfield) [reply packages/hash/frontend/src/components/hooks/useUpdateAuthenticatedUser.ts]: Does the variable even need to be optional then? Just wondering if we can catch it earlier and not need the assertion

File: packages/hash/frontend/src/pages/signup.page.tsx
Comment (Alfred-Mountfield) [reply packages/hash/frontend/src/pages/signup.page.tsx:222]: But if `errors: []` then it'd be falsey?

File: packages/hash/frontend/src/components/hooks/useUpdateAuthenticatedUser.ts
Comment (benwerner01) [reply packages/hash/frontend/src/components/hooks/useUpdateAuthenticatedUser.ts]: I can't think of a clean solution, as `kratosSession` has to be `undefined` at first render because it is fetched asynchronously. So in this function I can throw if its `undefined`, return if its `undefined`, but that would be dead code because as I said the queries will fail first.

File: packages/hash/frontend/src/pages/signup.page.tsx
Comment (benwerner01) [reply packages/hash/frontend/src/pages/signup.page.tsx:222]: No, `[]` is truthy
image

Comment (Alfred-Mountfield) [reply packages/hash/frontend/src/pages/signup.page.tsx:222]: image
right 🙃 

I hate JS.

File: packages/hash/frontend/src/components/hooks/useUpdateAuthenticatedUser.ts
Comment (Alfred-Mountfield) [reply packages/hash/frontend/src/components/hooks/useUpdateAuthenticatedUser.ts]: As discussed in [this thread](https://github.com/hashintel/hash/pull/1299#discussion_r1010743968) maybe we want to bring the check back?

Comment (benwerner01) [reply packages/hash/frontend/src/components/hooks/useUpdateAuthenticatedUser.ts]: Reverted in https://github.com/hashintel/hash/pull/1299/commits/80515e3be5a715ec14cb69e76d4066f9d8842d85

File: packages/hash/api/src/graphql/resolvers/knowledge/entity/before-update-entity-hooks.ts
Comment (benwerner01) [reply packages/hash/api/src/graphql/resolvers/knowledge/entity/before-update-entity-hooks.ts:38]: > just saying something like "invalid preferred name" is basically useless without explaining what condition was broken.

True, but the reason I'm doing this with the preferred name is so that its validation is consistent with the shortname validation. And if you look at `validateUserShortname`, you'll see that the error messages are more descriptive.

> We need to improve our error handling all over the place it's really not ideal right now. 

I agree, but coming up with a consistent approach which works across the various layers seems out of the scope of this PR to me. I'd prefer addressing this in follow-up, and we'd probably benefit from an internal discussion before I do.

Comment (Alfred-Mountfield) [reply packages/hash/api/src/graphql/resolvers/knowledge/entity/before-update-entity-hooks.ts:38]: Yes but as you say `validateAccountShortname` actually does report useful error information, so I don't think this is consistent.

Why don't we change `UserModel.preferredNameIsInvalid` to throw normal errors and then re-throw as the appropriate error here like you said? It doesn't seem like a lot of work

File: packages/hash/api/src/graphql/resolvers/knowledge/entity/before-update-entity-hooks.ts
Comment (benwerner01) [reply packages/hash/api/src/graphql/resolvers/knowledge/entity/before-update-entity-hooks.ts]: This is an old todo, yes - have created [the task](https://app.asana.com/0/1201095311341924/1203285346775714/f) and added a link to the asana task in https://github.com/hashintel/hash/pull/1299/commits/4c9d7a48ec7ae6c086597595dea2260ae01920ef

File: packages/hash/api/src/graphql/resolvers/knowledge/entity/before-update-entity-hooks.ts
Comment (benwerner01) [reply packages/hash/api/src/graphql/resolvers/knowledge/entity/before-update-entity-hooks.ts:41]: > Why don't we change UserModel.preferredNameIsInvalid to throw normal errors and then re-throw as the appropriate error here like you said? It doesn't seem like a lot of work

I think this won't be a simple pattern to get right. We will want to catch all errors, and then re-throw some errors as custom apollo errors with custom codes/messages, and re-throw anything unexpected as `INTERNAL_SERVER_ERROR`s. We should also introduce helper methods to be more consistent with how this works, and if we're already taking the time to refactor how we're throwing things in GraphQL it probably makes sense to get thoughts from FE reviewers. Therefore a standalone PR makes more sense to me, rather than addressing this now.

Comment (Alfred-Mountfield) [reply packages/hash/api/src/graphql/resolvers/knowledge/entity/before-update-entity-hooks.ts:41]: Right but I'm not suggesting we completely rework how we throw ApolloErrors. I'm just suggesting we throw errors from `preferredNameIsInvalid`, that's actually a completely separate concern from GraphQL and Apollo.

I appreciate what you're saying about knowing what to catch though, and in this case it'd just be one thing so I don't really think that's a problem..

If you really don't like the idea of doing it now that's fine, I just don't know if we're really going to catch all of these thoughts in some big task like "rethink errors in GraphQL". It'd be too easy to just take the bool here and throw a different kind of GraphQL error, not even ever thinking about looking at how `UserModel,preferredNameIsInvalid` behaves

File: packages/hash/api/src/graphql/resolvers/index.ts
Comment (benwerner01) [reply packages/hash/api/src/graphql/resolvers/index.ts:240]: Yeah, I'm not happy about this either - spent 5 minutes messing around with the `loggedInAndSignedUp` middleware method and deemed it'd be easiest to understand if the logic is captured in the resolver itself, although I hope we can come up with a better approach down the line.

File: packages/hash/api/src/graphql/resolvers/knowledge/entity/before-update-entity-hooks.ts
Comment (benwerner01) [reply packages/hash/api/src/graphql/resolvers/knowledge/entity/before-update-entity-hooks.ts]: Have refactored how the preferred name is validated in https://github.com/hashintel/hash/pull/1299/commits/ac23ce59bf0cf644bb90ec8c7d04c474d35fc5d7

File: packages/hash/frontend/src/components/hooks/useUpdateAuthenticatedUser.ts
Comment (benwerner01) [reply packages/hash/frontend/src/components/hooks/useUpdateAuthenticatedUser.ts]: Addressed as part of https://github.com/hashintel/hash/pull/1299/commits/4b12cf69722bd8000c03cab94332efec4b5dbaad

File: packages/hash/frontend/src/components/hooks/useUpdateAuthenticatedUser.ts
Comment (Alfred-Mountfield) [reply packages/hash/frontend/src/components/hooks/useUpdateAuthenticatedUser.ts]: I think I wasn't sure, or had forgotten, that we had a method called `extractEntityRoot` 😁 

My suggestion for `getRootsAsEntities` was just that it would handle all of the mapping and extraction of the inner of the vertex.

`extractEntityRoot` existing seems better, thanks

File: packages/hash/frontend/src/components/hooks/useUpdateAuthenticatedUser.ts
Comment (benwerner01) [reply packages/hash/frontend/src/components/hooks/useUpdateAuthenticatedUser.ts]: Have added the explicit check in https://github.com/hashintel/hash/pull/1299/commits/320831c20bbbefca558192427ea2d6d0fe4f2ed8.

> if we refactored the code here, re-ordered things, or changed to queries that don't require it, and TSC wouldn't warn us of that.

I'm not sure how this explicit check would let TSC warn us about anything in these scenarios, but I don't feel strongly and probably nice to have it be more explicit anyway.

File: packages/hash/frontend/src/components/hooks/useAuthenticatedUser.ts
Comment (benwerner01) [reply packages/hash/frontend/src/components/hooks/useAuthenticatedUser.ts:100]: > do we even need the me query to do what it does?

Interesting, haven't questioned this before. If the alternative is to have a `me` query that returns the `entityId` of the current user, and using the `getEntity` query to get the actual user entity, then we will end up having to do 2 consecutive requests where we previously only had to do one. This might be something worth optimising because the `me` query is fetched on every page-load to determine if the user is logged in, and obtain the latest user entity.

**Review by luisbettencourt — APPROVED** [review]
Thank for addressing the change I requested. Code looks good to me but you might want to chase an additional backend approval.

File: packages/hash/frontend/src/pages/signup.page.tsx
Comment (luisbettencourt) [reply packages/hash/frontend/src/pages/signup.page.tsx:222]: Isn't it still redundant, since you have the exact same check in `useUpdateAuthenticatedUser` before returning `errors`?

File: packages/hash/api/src/graphql/resolvers/knowledge/entity/before-update-entity-hooks.ts
Comment (thehabbos007) [inline packages/hash/api/src/graphql/resolvers/knowledge/entity/before-update-entity-hooks.ts:48]: Sorry if I'm missing something from the other reviews, but shouldn't this be fallible and allow cancellation of the rest of the entity saving chain? Or is the kratos communication handled somewhere else (for example)?

File: packages/hash/api/src/graphql/resolvers/knowledge/entity/entity.ts
Comment (thehabbos007) [inline packages/hash/api/src/graphql/resolvers/knowledge/entity/entity.ts:174]: I am assuming if we want the fallibility here, we would have to act on the callback, and only conditionally continue

File: packages/hash/api/src/graphql/resolvers/knowledge/entity/before-update-entity-hooks.ts
Comment (thehabbos007) [inline packages/hash/api/src/graphql/resolvers/knowledge/entity/before-update-entity-hooks.ts:80]: Is this the method that would handle side-effects? 
And if so, are we not tackling email address changes yet for kratos? :)

File: packages/hash/api/src/graphql/resolvers/user/updateUser.ts
Comment (thehabbos007) [inline packages/hash/api/src/graphql/resolvers/user/updateUser.ts:1]: Really nice we can get rid of these resolvers here!

**Review by thehabbos007 — COMMENTED** [review]
Hey, sorry for the late review. These things shouldn't be blocking the merge/approval, but I thought i should ask anyway. Thank you Ben! :)

File: packages/hash/api/src/graphql/resolvers/knowledge/entity/before-update-entity-hooks.ts
Comment (benwerner01) [reply packages/hash/api/src/graphql/resolvers/knowledge/entity/before-update-entity-hooks.ts:48]: > shouldn't this be fallible and allow cancellation of the rest of the entity saving chain?

Yes, the way this would currently work is that an error would be thrown in the callback method which would ensure the properties aren't updated. Is there an alternative method for accomplishing this you had in mind?

File: packages/hash/api/src/graphql/resolvers/knowledge/entity/entity.ts
Comment (benwerner01) [reply packages/hash/api/src/graphql/resolvers/knowledge/entity/entity.ts:174]: I'm not sure I'm understanding, but if the callback fails the error is propagated into this scope, and therefore the function wouldn't continue. We don't have to enforce this manually in the current design.

File: packages/hash/frontend/src/pages/signup.page.tsx
Comment (Alfred-Mountfield) [reply packages/hash/frontend/src/pages/signup.page.tsx:222]: If we do have the exact same one then should we change the types to make sure its actually captured by TS?

Something like `[Error, ...Error[]]?` or whatever, I'd be wary of us removing it here and then refactoring could mean we don't check it before hand again and the whole `[] is truthy` thing will just be a big pain to discover.

File: packages/hash/frontend/src/components/hooks/useAuthenticatedUser.ts
Comment (Alfred-Mountfield) [reply packages/hash/frontend/src/components/hooks/useAuthenticatedUser.ts:100]: That makes sense to me :) I mean ideally 2 requests should be _**very**_ cheap haha, but we've also already got the code so leaving it seems fine.

**Review by Alfred-Mountfield — APPROVED** [review]
Thanks for addressing everything, looks good to me too! A few open questions that deserve discussion offline (separate to this PR) I think

File: packages/hash/api/src/graphql/resolvers/knowledge/entity/before-update-entity-hooks.ts
Comment (benwerner01) [reply packages/hash/api/src/graphql/resolvers/knowledge/entity/before-update-entity-hooks.ts:80]: > are we not tackling email address changes yet for kratos?

Not yet, no - because this PR seeks to deprecate the existing `updateUser` mutation, which did not support updating email addresses.

I've captured this in [a follow-up task](https://app.asana.com/0/0/1203285346775716/f) :)

File: packages/hash/api/src/graphql/resolvers/knowledge/entity/before-update-entity-hooks.ts
Comment (thehabbos007) [reply packages/hash/api/src/graphql/resolvers/knowledge/entity/before-update-entity-hooks.ts:48]: I was thinking that the method could be explicit about failing by its return value instead. But that isn't really much different from just throwing/catching, especially if we just throw it all the way to the Apollo handler.

Being explicit may also allow us to get rid of the dependency on `"apollo-server-express"`. But we can accomplish that by defining our own exceptions. Not sure any of these changes would bring much value, though. 

Perhaps it could allow us to retry rather than bail in the `updateProperties` method if we have explicit failures? But unsure we need that right now.

File: packages/hash/api/src/graphql/resolvers/knowledge/entity/before-update-entity-hooks.ts
Comment (thehabbos007) [reply packages/hash/api/src/graphql/resolvers/knowledge/entity/before-update-entity-hooks.ts:80]: Thank you!

File: packages/hash/api/src/graphql/resolvers/knowledge/entity/before-update-entity-hooks.ts
Comment (benwerner01) [reply packages/hash/api/src/graphql/resolvers/knowledge/entity/before-update-entity-hooks.ts:48]: > Being explicit may also allow us to get rid of the dependency on "apollo-server-express". But we can accomplish that by defining our own exceptions. Not sure any of these changes would bring much value, though.

These methods are specifically for GraphQL resolvers, so I'm not too concerned by this dependency. If we want to move away from apollo errors, wouldn't we would lose some of the GQL error handling functionality and I'm not sure what we would be accomplishing... but there's lots of room for improvement regarding how we currently throw errors, so definitely expect things to change.

File: packages/hash/frontend/src/components/hooks/useAuthenticatedUser.ts
Comment (benwerner01) [reply packages/hash/frontend/src/components/hooks/useAuthenticatedUser.ts:100]: > ideally 2 requests should be very cheap

Server-side these operations are cheap, yes. But because these are browser requests the request time will largely depend on the client's internet connection. Therefore having 2 requests instead of 1 could have a noticeable impact on loading times.

**Review by thehabbos007 — APPROVED** [review]
Thanks, Ben! This is great :)

**Pull request merged**
